pkg/storage/chunk/client/azure/blob_storage_client.go

```diff
@@ -378,6 +382,37 @@ func (b *BlobStorage) getServicePrincipalToken() (*adal.ServicePrincipalToken, e
 
 	resource := fmt.Sprintf("https://%s.%s", b.cfg.StorageAccountName, endpoint)
 
+	if b.cfg.UseFederatedToken {
+		var environmentName string
+		if b.cfg.Environment == azureGlobal {
+			environmentName = "AzurePublicCloud"
+		} else {
+			environmentName = b.cfg.Environment
+		}
```

Comment:
Can you add `AzurePublicCloud` as part of the list of const `Environments` top of this file.

Also you could avoid `else` here. 
```
		environmentName := azurePublicCloud

		if b.cfg.Environment != azureGlobal {
			environmentName = b.cfg.Environment
		}

```
looks more readable IMHO.

Q: Also curious what's the difference between `AzureGlobal` vs `AzurePublicCloud` environments?

Reply:
I don't know where `AzureGlobal` is coming from but it doesn't exist in the Azure Go SDKs. It's `AzurePublicCloud` there, i.e. you cannot do an environment lookup via `EnvironmentFromName` with `AzureGlobal`

Reply:
<img width="682" alt="image" src="https://user-images.githubusercontent.com/1560587/199224507-54696b75-3bc3-4136-9096-bb7726c990ba.png">

The workload identity webhook is exposing a authority host `AZURE_AUTHORITY_HOST` which is the `env.ActiveDirectoryEndpoint`. Consider to prefer the environment variable instead a own cloud env lookup

Reply:
I thought about that but decided to go with the reference implementation outlined by the person who added the code to adal. I also looked at the new [azidentity module where AZURE_AUTHORITY_HOST gets only used as a fallback option.](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/azidentity.go#L79) It will only honor it if no cloud configuration has been set. However, the current implementation of Azure storage in Loki requires the set cloud environment name to be set, i.e. adding fallback logic is imho futile.